pynicotine/shares.py

```diff
@@ -833,15 +857,129 @@ def process_scanner_messages(self, scanner, scanner_queue):
 
         return False
 
-    def rescan_shares(self, init=False, rescan=True, rebuild=False, use_thread=True):
+    def list_shares(self, group=None):
+        """ Returns ONE string with all readable shares and missing shares.
+        group specifies which shares to list: "normal", "buddy", etc. """
+        # TODO: Unused, need new command in core_commands to run this
+
+        reads, fails = self.check_shares(shares=None, group=group)
+        num_reads = len(reads)
+        num_fails = len(fails)
+        num_total = num_reads + num_fails
+        total_shares_line = "0 " + _("shares") + " " + _("configured")
+
+        if not num_total:
+            fails.insert(0, total_shares_line)
+
+        # Coalesce the summary line toegether with all "Ready " and "Missing " items
+        all_shares = total_shares_line + ":\n\n" + '\n\n'.join(reads) + "\n" + '\n\n'.join(fails)
+
+        return all_shares
+
+    def check_shares(self, shares=None, group=None):
+        """ Returns TWO lists of strings: readable shares, and missing shares.
+        group specifies which shares to check: "normal", "buddy", etc. """
+
+        if shares is None:
+            # Triggered by list shares command
+            shares = self.get_shared_folders()
+
+        group_index = self.get_group_index(group) if group is not None else 0
+        reads, fails = [], []
+
+        for table in shares:
+            group_name = self.get_group_name(group_index)
+
+            for virtual, folder, *_unused in table:
+                if os.access(folder, os.R_OK):
+                    reads.append(_(f"Ready {group_name} share \"{virtual}\" at:\n{folder}"))
+                else:
+                    fails.append(_(f"Missing {group_name} share \"{virtual}\" at:\n{folder}"))
+
+            if group is not None:
+                # Only check a specific group (ie just "normal" or "buddy" shares, etc)
+                break
+
+            group_index += 1
+
+        return reads, fails
+
+    def confirm_force_rescan(self, shares):
+        """ Check if any shares are missing, if so then show a prompt """
+
+        reads, fails = self.check_shares(shares)
+
+        num_reads = len(reads)
+        num_fails = len(fails)
+        num_total = num_reads + num_fails
+
+        reads_shares_word = _("share") if num_reads == 1 else _("shares")
+        fails_shares_word = _("share") if num_fails == 1 else _("shares")
+        total_shares_word = _("share") if num_total == 1 else _("shares")
+
+        total_shares_line = f"{num_total} {total_shares_word} " + _("configured")
```

Comment:
The largest issue with this kind of string building is that the word order will be different in different languages. It's difficult for translators to a) realize that these strings should be translated in the context of a sentence, and b) piece them together. There could also be conflicts with individual words elsewhere in the application in the future.